x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingDeciderService.java

```diff
@@ -0,0 +1,561 @@
+                .mapToLong(this::getAnomalyMemoryRequirement)
+                .max()
+                .orElse(0L),
+            dataframeAnalyticsTasks.stream()
+                .filter(PersistentTask::isAssigned)
+                // Memory SHOULD be recently refreshed, so in our current state, we should at least have an idea of the memory used
+                .mapToLong(this::getAnalyticsMemoryRequirement)
+                .max()
+                .orElse(0L));
+
+        final Optional<AutoscalingDeciderResult> scaleDownDecision =
+            checkForScaleDown(nodes, clusterState, largestJob, currentScale, reasonBuilder);
+
+        if (scaleDownDecision.isPresent()) {
+            if (newScaleDownCheck()) {
+                scaleDownDetected = timeSupplier.get();
+            }
+            if (canScaleDown(decider.getDownScaleDelay())) {
+                return scaleDownDecision.get();
+            }
+            return new AutoscalingDeciderResult(
+                context.currentCapacity(),
+                reasonBuilder
+                    .setSimpleReason(
+                        "Passing currently perceived capacity as configured down scale delay has not be satisfied; configured delay ["
+                            + decider.getDownScaleDelay().millis()
+                            + "] last detected scale down event ["
+                            + scaleDownDetected
+                            + "]")
+                    .build());
+        }
+
+        return new AutoscalingDeciderResult(context.currentCapacity(),
+            reasonBuilder
+                .setSimpleReason("Passing currently perceived capacity as no scaling changes were detected to be possible")
+                .build());
+    }
+
+    Optional<AutoscalingDeciderResult> checkForScaleUp(MlAutoscalingDeciderConfiguration decider,
+                                                       List<String> waitingAnomalyJobs,
+                                                       List<String> waitingAnalyticsJobs,
+                                                       @Nullable NativeMemoryCapacity futureFreedCapacity,
+                                                       NativeMemoryCapacity currentScale,
+                                                       MlScalingReason.Builder reasonBuilder) {
+
+        // Are we in breach of maximum waiting jobs?
+        if (waitingAnalyticsJobs.size() > decider.getNumAnalyticsJobsInQueue()
+            || waitingAnomalyJobs.size() > decider.getNumAnomalyJobsInQueue()) {
+            NativeMemoryCapacity updatedCapacity = NativeMemoryCapacity.from(currentScale);
+            Optional<NativeMemoryCapacity> analyticsCapacity = requiredCapacityForUnassignedJobs(waitingAnalyticsJobs,
+                this::getAnalyticsMemoryRequirement,
+                // TODO Better default???
+                AnalysisLimits.DEFAULT_MODEL_MEMORY_LIMIT_MB,
```

Comment:
I think this is dangerous.  It means we could scale the cluster significantly due to the memory tracker not having the required information.  It would also be difficult to diagnose this based on a user complaint that their cluster scaled wildly.  For example, suppose there are 50 jobs each with a memory requirement of 0.1GB, total requirement 5GB, but then due to a glitch in the memory tracker we treat that as 50GB and scale up to a much more costly cluster.

Since this is for scaling up, I think we should use 0 for jobs where we have no memory information, so we'll only scale up if the sum of the memory requirements for jobs that we know the memory requirement for imply a scale up.

(Same a few lines below.)

Reply:
> Since this is for scaling up, I think we should use 0 for jobs where we have no memory information

The reason I did do this was this caused a scale_up to be delayed. If we are fine with the scale up being delayed, then its fine.

Reply:
I think this should only happen once per master node though, if the Cloud infrastructure asks for a scaling decision immediately after that master node became master.  In that rare situation it might be better not to make a hasty decision that will need to get adjusted again at the next scaling decision.

When a job is opened/started via the open/start API, we put a value for its memory requirement in the memory tracker.  And when a new master takes over, we recalculate the memory requirement of all jobs.

So the memory tracker should only return null for a job’s memory requirement in the period between a new master being elected and the recalculation of all jobs’ memory requirements completing.

I think it would be good to log a warning if the default is used during a scale up decision.  We'll need this to debug the situation where a cluster doesn't scale at all due to some unforeseen problem.  But unless I am mistaken I think it will be very rare to see that log message.